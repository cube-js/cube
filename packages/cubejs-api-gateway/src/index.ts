export * from './gateway';
export * from './interfaces';
export * from './CubejsHandlerError';
export * from './UserError';
export { getRequestIdFromRequest } from './requestParser';
