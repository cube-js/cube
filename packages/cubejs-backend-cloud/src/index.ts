export * from './cloud';
export * from './deploy';
