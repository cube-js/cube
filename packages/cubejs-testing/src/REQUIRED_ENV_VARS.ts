/**
 * Environment variables that cannot be hardcoded in code, and instead must be specified via the cli.
 * Usually cloud db config & auth variables.
 */
export const REQUIRED_ENV_VARS: {[key: string]: string[]} = {
  athena: [
    'CUBEJS_AWS_KEY',
    'CUBEJS_AWS_SECRET',
    'CUBEJS_AWS_REGION',
    'CUBEJS_AWS_S3_OUTPUT_LOCATION',
    'CUBEJS_DB_EXPORT_BUCKET'
  ],
  bigquery: [
    'CUBEJS_DB_BQ_PROJECT_ID',
    'CUBEJS_DB_EXPORT_BUCKET',
    'CUBEJS_DB_BQ_CREDENTIALS',
  ],
  firebolt: [
    'CUBEJS_DB_USER',
    'CUBEJS_DB_PASS',
    'CUBEJS_DB_NAME',
    'CUBEJS_FIREBOLT_ENGINE_ENDPOINT'
  ],
  materialize: [],
  multidb: [],
  questdb: [],
  postgres: [],
  redshift: [
    'CUBEJS_DB_HOST',
    'CUBEJS_DB_PORT',
    'CUBEJS_DB_NAME',
    'CUBEJS_DB_USER',
    'CUBEJS_DB_PASS',
  ],
};
