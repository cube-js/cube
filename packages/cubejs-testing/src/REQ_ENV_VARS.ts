/**
 * Required environment variables per datasource.
 */
export const REQ_ENV_VARS: {[key: string]: string[]} = {
  athena: [
    'CUBEJS_AWS_KEY',
    'CUBEJS_AWS_SECRET',
    'CUBEJS_AWS_REGION',
    'CUBEJS_AWS_S3_OUTPUT_LOCATION',
    'CUBEJS_DB_EXPORT_BUCKET'
  ],
  bigquery: [
    'CUBEJS_DB_BQ_PROJECT_ID',
    'CUBEJS_DB_EXPORT_BUCKET',
    'CUBEJS_DB_BQ_CREDENTIALS',
  ],
  postgres: [
    'CUBEJS_DB_USER',
    'CUBEJS_DB_PASS',
    'CUBEJS_DB_TYPE',
  ],
  questdb: [],
};
