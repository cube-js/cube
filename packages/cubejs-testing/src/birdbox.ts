import path from 'path';
import fs from 'fs';
import { spawn } from 'child_process';
import HttpProxy from 'http-proxy';
import { DockerComposeEnvironment, StartedTestContainer } from 'testcontainers';
import { execInDir, pausePromise } from '@cubejs-backend/shared';
import { getLocalHostnameByOs, PostgresDBRunner } from '@cubejs-backend/testing-shared';
import fsExtra from 'fs-extra';

export interface BirdBoxTestCaseOptions {
  name: string;
  loadScript?: string;
  env?: Record<string, string>;
}

export interface BirdBox {
  stop: () => Promise<void>;
  configuration: {
    playgroundUrl: string;
    apiUrl: string;
    wsUrl: string;
    env?: Record<string, string>;
  };
}

export async function startBirdBoxFromContainer(options: BirdBoxTestCaseOptions): Promise<BirdBox> {
  if (process.env.TEST_CUBE_HOST) {
    const host = process.env.TEST_CUBE_HOST || 'localhost';
    const port = process.env.TEST_CUBE_PORT || '8888';

    return {
      stop: async () => {
        console.log('[Birdbox] Closed');
      },
      configuration: {
        playgroundUrl: `http://${host}:${port}`,
        apiUrl: `http://${host}:${port}/cubejs-api/v1`,
        wsUrl: `ws://${host}:${port}`,
      },
    };
  }

  if (process.env.BIRDBOX_CUBEJS_REGISTRY_PATH === undefined) {
    process.env.BIRDBOX_CUBEJS_REGISTRY_PATH = 'localhost:5000/';
  }

  if (process.env.BIRDBOX_CUBEJS_VERSION === undefined) {
    process.env.BIRDBOX_CUBEJS_VERSION = 'latest';
    const tag = `${process.env.BIRDBOX_CUBEJS_REGISTRY_PATH}cubejs/cube:${process.env.BIRDBOX_CUBEJS_VERSION}`;
    if (execInDir('../..', `docker build . -f packages/cubejs-docker/dev.Dockerfile -t ${tag}`) !== 0) {
      throw new Error('[Birdbox] Docker build failed.');
    }
  }

  if (process.env.BIRDBOX_CUBESTORE_VERSION === undefined) {
    process.env.BIRDBOX_CUBESTORE_VERSION = 'latest';
  }

  const composeFile = `${options.name}.yml`;
  let dc = new DockerComposeEnvironment(
    path.resolve(path.dirname(__filename), '../../birdbox-fixtures/'),
    composeFile
  );

  if (options.env) {
    for (const k of Object.keys(options.env)) {
      dc = dc.withEnv(k, options.env[k]);
    }
  }

  console.log(`[Birdbox] Using ${composeFile} compose file`);

  const env = await dc
    .withStartupTimeout(30 * 1000)
    .withEnv('BIRDBOX_CUBEJS_VERSION', process.env.BIRDBOX_CUBEJS_VERSION)
    .withEnv('BIRDBOX_CUBESTORE_VERSION', process.env.BIRDBOX_CUBESTORE_VERSION)
    .up();

  const host = '127.0.0.1';
  const port = env.getContainer('birdbox-cube').getMappedPort(4000);
  const playgroundPort = process.env.TEST_PLAYGROUND_PORT ?? port;

  let proxyServer: HttpProxy | null = null;

  if (process.env.TEST_PLAYGROUND_PORT) {
    console.log(`[Birdbox] Creating a proxy server 4000->${port} for local testing`);
    // As local Playground proxies requests to the 4000 port
    proxyServer = HttpProxy.createProxyServer({ target: `http://localhost:${port}` }).listen(4000);

    proxyServer.on('error', async (err, req, res) => {
      console.log('[Proxy Server] error:', err);

      if (!res.headersSent) {
        res.writeHead(500, { 'content-type': 'application/json' });
      }

      res.end(JSON.stringify({ error: err.message }));
    });
  }

  if (options.loadScript) {
    const { loadScript } = options;
    console.log(`[Birdbox] Executing ${loadScript} script`);

    const { output, exitCode } = await env.getContainer('birdbox-db').exec([`/scripts/${loadScript}`]);

    if (exitCode === 0) {
      console.log(`[Birdbox] Script ${loadScript} finished successfully`);
    } else {
      console.log(output);

      console.log(`[Birdbox] Script ${loadScript} finished with error: ${exitCode}`);

      await env.down();

      process.exit(1);
    }
  }

  return {
    stop: async () => {
      console.log('[Birdbox] Closing');

      await env.down();
      proxyServer?.close();

      console.log('[Birdbox] Closed');
    },
    configuration: {
      playgroundUrl: `http://${host}:${playgroundPort}`,
      apiUrl: `http://${host}:${port}/cubejs-api/v1`,
      wsUrl: `ws://${host}:${port}`,
      env: {
        ...(process.env.TEST_PLAYGROUND_PORT ? { CUBEJS_DB_HOST: getLocalHostnameByOs() } : null),
      },
    },
  };
}

export interface StartCliWithEnvOptions {
  dbType: string;
  useCubejsServerBinary?: boolean;
  loadScript?: string;
  cubejsConfig?: string;
  env?: Record<string, string>;
}

export async function startBirdBoxFromCli(options: StartCliWithEnvOptions): Promise<BirdBox> {
  let db: StartedTestContainer;

  if (options.loadScript) {
    db = await PostgresDBRunner.startContainer({
      volumes: [
        {
          source: path.join(__dirname, '..', '..', 'birdbox-fixtures', 'datasets'),
          target: '/data',
          bindMode: 'ro',
        },
        {
          source: path.join(__dirname, '..', '..', 'birdbox-fixtures', 'postgresql', 'scripts'),
          target: '/scripts',
          bindMode: 'ro',
        },
      ],
    });

    console.log('[Birdbox] Executing load script');

    const loadScript = `/scripts/${options.loadScript}`;
    const { output, exitCode } = await db.exec([loadScript]);

    if (exitCode === 0) {
      console.log(`[Birdbox] Script ${loadScript} finished successfully`);
    } else {
      console.log(output);

      console.log(`[Birdbox] Script ${loadScript} finished with error: ${exitCode}`);

      await db.stop();

      process.exit(1);
    }
  }

  const testDir = path.join(process.cwd(), 'birdbox-test-project');

  // Do not remove whole dir as it contains node_modules
  if (fs.existsSync(path.join(testDir, '.env'))) {
    fs.unlinkSync(path.join(testDir, '.env'));
  }

  if (fs.existsSync(path.join(testDir, '.cubestore'))) {
    fsExtra.removeSync(path.join(testDir, '.cubestore'));
  }

  fsExtra.copySync(
    path.join(process.cwd(), 'birdbox-fixtures', 'postgresql'),
    path.join(testDir)
  );

  if (options.cubejsConfig) {
    fsExtra.copySync(
      path.join(process.cwd(), 'birdbox-fixtures', 'postgresql', options.cubejsConfig),
      path.join(testDir, 'cube.js')
    );
  }

  const cli = spawn(
    options.useCubejsServerBinary ? path.resolve(process.cwd(), '../cubejs-server/bin/server') : 'npm',
    options.useCubejsServerBinary ? [] : ['run', 'dev'],
    {
      cwd: testDir,
      shell: true,
      // Show output of Cube.js process in console
      stdio: ['pipe', 'pipe', 'pipe'],
      env: {
        ...process.env,
        CUBEJS_DB_TYPE: options.dbType === 'postgresql' ? 'postgres' : options.dbType,
        CUBEJS_DEV_MODE: 'true',
        CUBEJS_API_SECRET: 'mysupersecret',
        CUBEJS_WEB_SOCKETS: 'true',
        CUBEJS_PLAYGROUND_AUTH_SECRET: 'mysupersecret',
        ...options.env
          ? options.env
          : {
            CUBEJS_DB_HOST: db!.getHost(),
            CUBEJS_DB_PORT: `${db!.getMappedPort(5432)}`,
            CUBEJS_DB_NAME: 'test',
            CUBEJS_DB_USER: 'test',
            CUBEJS_DB_PASS: 'test',
          }
      },
    }
  );
  cli.stdout.on('data', (msg) => {
    console.log(msg.toString());
  });
  cli.stderr.on('data', (msg) => {
    console.log(msg.toString());
  });
  // cli.on('exit', (code) => {
  //   throw new Error(
  //     `Cube.js exited with ${code}`
  //   );
  // });

  await pausePromise(10 * 1000);

  return {
    stop: async () => {
      console.log('[Birdbox] Closing');

      if (db) {
        await db.stop();
      }

      console.log('[Birdbox] Done with DB');

      cli.kill();

      console.log('[Birdbox] Closed');
    },
    configuration: {
      playgroundUrl: 'http://127.0.0.1:4000',
      apiUrl: 'http://127.0.0.1:4000/cubejs-api/v1',
      wsUrl: 'ws://127.0.0.1:4000',
    },
  };
}
