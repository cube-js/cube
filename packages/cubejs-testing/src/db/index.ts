export * from './mysql';
export * from './postgres';
export * from './cubestore';
