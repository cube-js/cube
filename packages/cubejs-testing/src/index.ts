export * from './birdbox';
export * from './db';
export * from './query-test.abstract';
