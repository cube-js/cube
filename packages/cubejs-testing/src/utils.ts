import { platform } from 'os';

export function getLocalHostnameByOs() {
  if (platform() === 'win32') {
    return 'docker.for.win.localhost';
  }

  if (platform() === 'darwin') {
    return 'host.docker.internal';
  }

  return 'localhost';
}
