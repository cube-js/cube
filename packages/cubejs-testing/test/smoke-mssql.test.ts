import cubejs, { CubejsApi } from '@cubejs-client/core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { afterAll, beforeAll, expect, jest } from '@jest/globals';
import { MssqlDbRunner } from '@cubejs-backend/testing-shared';
import { BirdBox, getBirdbox } from '../src';
import { DEFAULT_API_TOKEN, DEFAULT_CONFIG, testQueryMeasure } from './smoke-tests';

describe('mssql', () => {
  jest.setTimeout(60 * 5 * 1000);
  let birdbox: BirdBox;
  let client: CubejsApi;

  beforeAll(async () => {
    const db = await MssqlDbRunner.startContainer({});
    birdbox = await getBirdbox(
      'mssql',
      {
        CUBEJS_DB_TYPE: 'mssql',

        CUBEJS_DB_HOST: db.getHost(),
        CUBEJS_DB_PORT: `${db.getMappedPort(1433)}`,
        CUBEJS_DB_USER: 'sa',
        CUBEJS_DB_PASS: process.env.TEST_DB_PASSWORD || 'Test1test',

        ...DEFAULT_CONFIG,
      },
      {
        schemaDir: 'mssql/schema',
      }
    );
    client = cubejs(async () => DEFAULT_API_TOKEN, {
      apiUrl: birdbox.configuration.apiUrl,
    });
  });

  afterAll(async () => {
    await birdbox.stop();
  });

  test('query measure', () => testQueryMeasure(client));

  test('query dimensions', async () => {
    const response = await client.load({
      measures: [
        'Orders.totalAmount',
      ],
      dimensions: [
        'Orders.status',
      ],
    });

    expect(response.rawData()).toMatchSnapshot('dimensions');
  });
});
