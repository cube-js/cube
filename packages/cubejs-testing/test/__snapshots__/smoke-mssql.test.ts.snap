// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mssql query dimensions: dimensions 1`] = `
Array [
  Object {
    "Orders.status": "processed",
    "Orders.totalAmount": "800",
  },
  Object {
    "Orders.status": "cancelled_by_customer",
    "Orders.totalAmount": "700",
  },
  Object {
    "Orders.status": "shipped",
    "Orders.totalAmount": "600",
  },
  Object {
    "Orders.status": "new",
    "Orders.totalAmount": "300",
  },
]
`;

exports[`mssql query measure: query 1`] = `
Array [
  Object {
    "Orders.totalAmount": "2400",
  },
]
`;
